buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java

```diff
@@ -1940,6 +1945,642 @@ public void testNioBufferExposeOnlyRegion() {
+        assertEquals(1, buf.refCnt());
+        assertTrue(buf.duplicate().release());
+        assertEquals(0, buf.refCnt());
+    }
+
+    // Test-case trying to reproduce:
+    // https://github.com/netty/netty/issues/2843
+    @Test
+    public void testRefCnt() throws Exception {
+        testRefCnt0(false);
+    }
+
+    // Test-case trying to reproduce:
+    // https://github.com/netty/netty/issues/2843
+    @Test
+    public void testRefCnt2() throws Exception {
+        testRefCnt0(true);
+    }
+
+    @Test
+    public void testEmptyNioBuffers() throws Exception {
+        ByteBuf buffer = releaseLater(newBuffer(8));
+        buffer.clear();
+        assertFalse(buffer.isReadable());
+        ByteBuffer[] nioBuffers = buffer.nioBuffers();
+        assertEquals(1, nioBuffers.length);
+        assertFalse(nioBuffers[0].hasRemaining());
+    }
+
+    @Test
+    public void testGetReadOnlyDirectDst() {
+        testGetReadOnlyDst(true);
+    }
+
+    @Test
+    public void testGetReadOnlyHeapDst() {
+        testGetReadOnlyDst(false);
+    }
+
+    private void testGetReadOnlyDst(boolean direct) {
+        byte[] bytes = { 'a', 'b', 'c', 'd' };
+
+        ByteBuf buffer = releaseLater(newBuffer(bytes.length));
+        buffer.writeBytes(bytes);
+
+        ByteBuffer dst = direct ? ByteBuffer.allocateDirect(bytes.length) : ByteBuffer.allocate(bytes.length);
+        ByteBuffer readOnlyDst = dst.asReadOnlyBuffer();
+        try {
+            buffer.getBytes(0, readOnlyDst);
+            fail();
+        } catch (ReadOnlyBufferException e) {
+            // expected
+        }
+        assertEquals(0, readOnlyDst.position());
+    }
+
+    @Test
+    public void testReadBytes() {
+        ByteBuf buffer = newBuffer(8);
+        byte[] bytes = new byte[8];
+        buffer.writeBytes(bytes);
+
+        ByteBuf buffer2 = buffer.readBytes(4);
+        assertSame(buffer.alloc(), buffer2.alloc());
+        assertEquals(4, buffer.readerIndex());
+        assertTrue(buffer.release());
+        assertEquals(0, buffer.refCnt());
+        assertTrue(buffer2.release());
+        assertEquals(0, buffer2.refCnt());
+    }
+
+    private void testRefCnt0(final boolean parameter) throws Exception {
+        for (int i = 0; i < 10; i++) {
+            final CountDownLatch latch = new CountDownLatch(1);
+            final CountDownLatch innerLatch = new CountDownLatch(1);
+
+            final ByteBuf buffer = newBuffer(4);
+            assertEquals(1, buffer.refCnt());
+            final AtomicInteger cnt = new AtomicInteger(Integer.MAX_VALUE);
+            Thread t1 = new Thread(new Runnable() {
+                @Override
+                public void run() {
+                    boolean released;
+                    if (parameter) {
+                        released = buffer.release(buffer.refCnt());
+                    } else {
+                        released = buffer.release();
+                    }
+                    assertTrue(released);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this assertion. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2186)